File: src/BootstrapBlazor.WebAPI/WebSpeech.razor.js
let speakBtn, el, utteranceCache, recognition;
let inited = false;
let speechSynthesis = window.speechSynthesis || window.mozspeechSynthesis || window.webkitspeechSynthesis;

export function SpeechRecognition(wrapper, lang) {

    if (!lang) lang = 'zh-CN';

    var SpeechRecognition = SpeechRecognition || webkitSpeechRecognition

    var colors = ['水色', '天蓝色', '米色', '素色', '黑色', '蓝色', '棕色', '巧克力', '珊瑚', '深红色', '青色', '紫红色', '鬼白', '金色', '金黄色', '灰色', '绿色', '靛蓝', '象牙色', '卡其色', '薰衣草色', '石灰', '亚麻色', '洋红色', '栗色', '莫卡辛', '海军蓝', '橄榄色', '橙色', '兰花', '秘鲁', '粉色', '梅花', '紫色', '红色', '鲑鱼色', '西耶娜', '银色', '雪', '棕褐色', '青色', '蓟', '番茄', '绿松石', '紫罗兰', '白色', '黄色'];

    var colorsEn = ['aqua', 'azure', 'beige', 'bisque', 'black', 'blue', 'brown', 'chocolate', 'coral', 'crimson', 'cyan', 'fuchsia', 'ghostwhite', 'gold', 'goldenrod', 'gray', 'green', 'indigo', 'ivory', 'khaki', 'lavender', 'lime', 'linen', 'magenta', 'maroon', 'moccasin', 'navy', 'olive', 'orange', 'orchid', 'peru', 'pink', 'plum', 'purple', 'red', 'salmon', 'sienna', 'silver', 'snow', 'tan', 'teal', 'thistle', 'tomato', 'turquoise', 'violet', 'white', 'yellow'];

    recognition = new SpeechRecognition();
    recognition.continuous = false;
    recognition.lang = lang;
    recognition.interimResults = false;
    recognition.maxAlternatives = 1;

    var diagnostic = document.querySelector('.output');
    var hints = document.querySelector('.hints');

    var colorHTML = '';
    colors.forEach(function (v, i, a) {
        console.log(v, i);
        colorHTML += '<span style="background-color:' + colorsEn[i] + ';"> ' + v + ' </span>';
    });
    hints.innerHTML = '<h4>点击这个区域，然后说出一种颜色来更改应用程序的背景颜色<br/> ' + colorHTML + '.<h4>';

    hints.onclick = function () {
        if (recognition) {
            recognition.stop();
        }
        recognition.start();
        console.log('开始识别.');
        wrapper.invokeMethodAsync('GetStatus', "开始识别");
    }

    recognition.onresult = function (event) {
        var color = event.results[0][0].transcript;
        wrapper.invokeMethodAsync('GetResult', color);
        diagnostic.textContent = '结果: ' + color;
        var index = colors.indexOf(color.replace('.', '').replace('。', ''));
        if (index > -1) {
            color = colorsEn[index];
            diagnostic.textContent += '(' + color + ')';
        }
        hints.style.backgroundColor = color;
        diagnostic.textContent += ' , ' + '匹配率: ' + event.results[0][0].confidence;
        console.log('结果: ' + event.results[0][0].transcript + ',匹配率: ' + event.results[0][0].confidence);
        return event.results[0][0].transcript;
    }

    recognition.onspeechend = function () {
        recognition.stop();
    }

    recognition.onnomatch = function () {
        diagnostic.textContent = "不能识别.";
        wrapper.invokeMethodAsync('GetStatus', "不能识别");
    }

    recognition.onerror = function (event) {
        diagnostic.textContent = '识别时出现错误: ' + event.error;
        wrapper.invokeMethodAsync('GetError', event.error);
    }

}
export function SpeechRecognitionStop(wrapper) {
    try {
        if (recognition) {
            recognition.stop();
            wrapper.invokeMethodAsync('GetStatus', "停止识别");
        }
    } catch (error) {
        console.error(error);
        wrapper.invokeMethodAsync('GetError', error.message);
        return false;
    }
    return true;
}

export function SpeechSynthesis(wrapper, text, lang, rate = 1, picth = 1, volume = 1, voiceURI = null) {
    try {
        if (!wrapper) {
            return;
        }
        if (typeof speechSynthesis === "undefined") {
            wrapper.invokeMethodAsync('GetError', "Speech undefined");
            return;
        }
        if (!speechSynthesis) {
            wrapper.invokeMethodAsync('GetError', "Speech 未初始化");
        }
        if (!text) text = 'hello blazor';
        if (!lang) lang = 'zh-CN';

        // Create speech synthesis utterance
        var utterance = new SpeechSynthesisUtterance();

        // Speech synthesis voices 返回一个 SpeechSynthesisVoice 列表，用于表示当前设备上所有可用的语音。
        const voices = speechSynthesis.getVoices();

        // Set voice and language
        if (voiceURI !== null)
            utterance.voice = voices.find(voice => voice.voiceURI === voiceURI);
        else
            utterance.voice = voices.find(voice => voice.lang === lang);

        wrapper.invokeMethodAsync('GetStatus', '语音:' + utterance.voice.name);

        utterance.text = text;
        utterance.volume = volume;
        utterance.rate = rate;
        utterance.pitch = picth;

        // Speak the text
        speechSynthesis.speak(utterance);
        if (!inited) {
            setTimeout(() => {
                if (!speechSynthesis.speaking) {
                    wrapper.invokeMethodAsync('GetError', '请点击初始化按钮');
                    utteranceCache = utterance;
                    el.style.visibility = "visible";
                    speakBtn.addEventListener("click", talk, false);
                } else {
                    inited = true;
                }
            }, 100);
        }
    } catch (error) {
        console.error(error);
        wrapper.invokeMethodAsync('GetError', error.message);
    }
}

export function SpeechStop(wrapper) {
    try {
        if (speechSynthesis.speaking) {
            speechSynthesis.cancel();
            wrapper.invokeMethodAsync('GetStatus', "停止播放");
        }
    } catch (error) {
        console.error(error);
        wrapper.invokeMethodAsync('GetError', error.message);
        return false;
    }
    return true;
}

export async function GetVoiceList(wrapper) {
    if (typeof speechSynthesis === "undefined") {
        wrapper.invokeMethodAsync('GetError', "Speech undefined");
        return;
    }
    let voices = speechSynthesis.getVoices();
    let list = [];
    for (var i = 0; i < voices.length; i++) {
        let x = {
            name: voices[i].name,
            isDefault: voices[i].default,
            lang: voices[i].lang,
            localService: voices[i].localService,
            voiceURI: voices[i].voiceURI,
        }
        list.push(x)
    }
    return list;
}

export function InitWebapi(wrapper, element) {
    el = element;
    speakBtn = element.querySelector("[data-action=speakBtn]");
}

function talk() {
    //speechSynthesis.speak() without user activation is no longer allowed
    //iOS 禁止网站自动播放视频，需要用户交互才能播放。此策略已扩展到 voiceSynthesis.speak 调用。Chrome 也效仿，现在打印一条弃用警告：
    //自 M71（2018 年 12 月左右）起，不再允许在没有用户激活的情况下使用 voiceSynthesis.speak()。有关更多详细信息，请参阅 https://www.chromestatus.com/feature/5687444770914304
    el.style.visibility = "hidden";
    if (utteranceCache !== null) {
        speechSynthesis.speak(utteranceCache);
    } else {
        let u = new SpeechSynthesisUtterance();
        u.text = 'OK';
        speechSynthesis.speak(u);
    }
    setTimeout(() => {
        if (!speechSynthesis.speaking) {
            el.style.visibility = "visible";
            speakBtn.addEventListener("click", talk, false);
        } else {
            inited = true;
            speakBtn.removeEventListener("click", talk);
        }
    }, 100);
}

File: src/BootstrapBlazor.WebAPI/WebSpeech.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace BootstrapBlazor.Components;

/// <summary>
/// 
/// </summary>
public partial class WebSpeech : IAsyncDisposable
{
    [Inject] IJSRuntime? JS { get; set; }
    private IJSObjectReference? module;
    private DotNetObjectReference<WebSpeech>? Instance { get; set; }

    /// <summary>
    /// UI界面元素的引用对象
    /// </summary>
    public ElementReference Element { get; set; }

    /// <summary>
    /// 获得/设置 错误回调方法
    /// </summary>
    [Parameter]
    public Func<string, Task>? OnError { get; set; }

    /// <summary>
    /// 获得/设置 识别完成回调方法
    /// </summary>
    [Parameter]
    public Func<string, Task>? OnResult { get; set; }

    /// <summary>
    /// 获得/设置 状态信息回调方法
    /// </summary>
    [Parameter]
    public Func<string, Task>? OnStatus { get; set; }

    /// <summary>
    /// 显示语音识别调试信息
    /// </summary>
    [Parameter]
    public bool ShowSpeechInfo { get; set; }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        try
        {
            if (firstRender)
            {
                module = await JS!.InvokeAsync<IJSObjectReference>("import", "./_content/BootstrapBlazor.WebAPI/WebSpeech.razor.js" + "?v=" + System.Reflection.Assembly.GetExecutingAssembly().GetName().Version);
                Instance = DotNetObjectReference.Create(this);
                await InitWebapi();
            } 
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
        }
    }

    [JSInvokable]
    public async Task GetStatus(string val)
    {
        if (OnStatus != null) await OnStatus.Invoke(val);
    }

    [JSInvokable]
    public async Task GetResult(string val)
    {
        if (OnResult != null) await OnResult.Invoke(val);
    }

    /// <summary>
    /// 初始化语音
    /// </summary>
    public virtual async Task InitWebapi()
    {
        try
        {
            await module!.InvokeVoidAsync("InitWebapi", Instance, Element);
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
        }
    }

    /// <summary>
    /// 语音识别
    /// </summary>
    /// <returns></returns>
    public virtual async Task<string> SpeechRecognition(string lang = "zh-CN")
    {
        try
        {
            return await module!.InvokeAsync<string>("SpeechRecognition", Instance, lang);
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
            return e.Message;
        }
    }

    /// <summary>
    /// 语音合成（文字转语音）
    /// </summary>
    /// <param name="text">文字</param>
    /// <param name="lang">语言</param>
    /// <param name="rate">速率, 范围可以在 0.1（最低）和 10（最高）之间</param>
    /// <param name="picth">音高,范围可以在 0（最低）和 2（最高）之间</param>
    /// <param name="volume">音量, 浮点数，介于 0（最低）和 1（最高）之间</param>
    /// <param name="voiceURI">语音引擎名称</param>
    /// <returns></returns>
    public virtual async Task SpeechSynthesis(string text, string lang = "zh-CN", double rate = 1, double picth = 1, double volume = 1, string? voiceURI = null)
    {
        try
        {
            await module!.InvokeVoidAsync("SpeechSynthesis", Instance, text, lang, rate, picth, volume, voiceURI);
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
        }
    }

    public virtual async Task SpeechRecognitionStop()
    {
        try
        {
            var res = await module!.InvokeAsync<bool>("SpeechRecognitionStop", Instance);
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
        }
    }

    public virtual async Task SpeechStop()
    {
        try
        {
            var res = await module!.InvokeAsync<bool>("SpeechStop", Instance);
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke(e.Message);
        }
    }

    public virtual async Task<List<WebVoice>?> GetVoiceList(bool orderByName = false)
    {
        try
        {
            List<WebVoice> res = await module!.InvokeAsync<List<WebVoice>>("GetVoiceList");
            var retry = 0;
            while (res == null || res.Count == 0)
            {
                await Task.Delay(200);
                res = await module!.InvokeAsync<List<WebVoice>>("GetVoiceList", Instance);
                retry++;
                if (retry == 5)
                {
                    return null;
                }
            }
            try
            {
                return orderByName ? (res?.OrderByDescending(a => a.LocalService).ThenBy(a => a.Lang).ThenBy(a => a.Name).ToList()) : res;
            }
            catch (Exception)
            { 
                return res;
            }
        }
        catch (Exception e)
        {
            if (OnError != null) await OnError.Invoke("GetVoiceList:" + e.Message);
        }
        return null;
    }

    async ValueTask IAsyncDisposable.DisposeAsync()
    {
        Instance?.Dispose();
        if (module is not null)
        {
            await module.DisposeAsync();
        }
    }

    

}
